torch/optim/sparse_adam.py

```diff
@@ -97,11 +97,36 @@ def step(self, closure=None):
 
         return loss
 
-SparseAdam.__doc__ = r"""Implements lazy version of Adam algorithm suitable for sparse tensors.
-
-    In this variant, only moments that show up in the gradient get updated, and
-    only those portions of the gradient get applied to the parameters.
-
+SparseAdam.__doc__ = r"""SparseAdam implements a lazy version of the Adam algorithm
+    suitable for sparse gradients. Please note that SparseAdam is NOT equivalent to
+    Adam "but for sparse gradients", and we understand this may be confusing.
+
+    This is a semantically different optimizer where only moments that show up in the
+    gradient get updated, and only those portions of the gradient get applied to the
+    parameters. This is an optimization, since the 0-values in the gradient will be
+    skipped, and one can presume that this is only an optimization in semantically
+    sparse tensors, meaning there are lots of 0-values that are skipped. We implement
+    the optimization by using the sparse layout to do the update computations. This
+    does NOT mean that SparseAdam is especially meant for tensors with a sparse layout,
+    as some of our other optimizers can handle tensors with a sparse layout as well.
+
```

Comment:
It might be useful to give an example of when this would actually even occur. How one could achieve sparse gradients (where `param.grad.is_sparse` will return `True`) and dense parameters?

Reply:
I think I mention it later on--is the later comment sufficient?

Reply:
From looking through the embedding implementation in torch (which eventually dispatches to `embedding_sparse_backward`), the sparse tensors are created from keeping track of actual indices and values which get are needed to construct sparse (semantically and in layout) gradients.

Reply:
So instead of declaring a parameter using `nn.Parameter `, I could just declare it using `nn.Embedding(sparse=True)`?

Reply:
Oh no, nn.Embedding is an actual module that involves embedding bags and will do computation on your parameters. I don't really know what the implications of using it will be, but it is not a replacement for nn.Parameter.

My earlier answer is suggesting that you may want to rewrite your backward to directly initialize and construct sparse grads, but I don't anticipate this will be easy or trivial. If what you're looking for is a performance win, I would still recommend fused adam!